clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java

```diff
@@ -181,63 +348,560 @@ public void updateState(ConsumerGroupHeartbeatResponseData response) {
+            return Optional.ofNullable(nameFromSubscriptionCache);
         }
     }
 
     /**
-     * Returns true if the member has a target assignment being processed.
+     * Build set of TopicPartition for the partitions included in the heartbeat topicPartitions,
+     * and using the given topic name.
      */
-    private boolean hasPendingTargetAssignment() {
-        return targetAssignment.isPresent();
+    private SortedSet<TopicPartition> buildAssignedPartitionsWithTopicName(
+            List<Integer> topicPartitions,
+            String topicName) {
+        SortedSet<TopicPartition> assignedPartitions = new TreeSet<>(COMPARATOR);
+        topicPartitions.forEach(tp -> assignedPartitions.add(new TopicPartition(topicName, tp)));
+        return assignedPartitions;
+    }
+
+    /**
+     * Revoke partitions. This will:
+     * <ul>
+     *     <li>Trigger an async commit offsets request if auto-commit enabled.</li>
+     *     <li>Invoke the onPartitionsRevoked callback if the user has registered it.</li>
+     * </ul>
+     *
+     * This will wait on the commit request to finish before invoking the callback. If the commit
+     * request fails, this will proceed to invoke the user callbacks anyway,
+     * returning a future that will complete or fail depending on the callback execution only.
+     *
+     * @param revokedPartitions Partitions to revoke.
+     * @return Future that will complete when the commit request and user callback completes.
+     */
+    private CompletableFuture<Void> revokePartitions(Set<TopicPartition> revokedPartitions) {
+        log.info("Revoking previously assigned partitions {}", Utils.join(revokedPartitions, ", "));
+
+        logPausedPartitionsBeingRevoked(revokedPartitions);
+
+        // Mark partitions as pending revocation to stop fetching from the partitions (no new
+        // fetches sent out, and no in-flight fetches responses processed).
+        markPendingRevocationToPauseFetching(revokedPartitions);
+
+        // Future that will complete when the revocation completes (including offset commit
+        // request and user callback execution)
+        CompletableFuture<Void> revocationResult = new CompletableFuture<>();
+
+        // Commit offsets if auto-commit enabled.
+        CompletableFuture<Void> commitResult;
+        if (commitRequestManager.autoCommitEnabled()) {
+            commitResult = commitRequestManager.maybeAutoCommitAllConsumed();
+        } else {
+            commitResult = CompletableFuture.completedFuture(null);
+        }
+
+        commitResult.whenComplete((result, error) -> {
+
+            if (error != null) {
+                // Commit request failed (commit request manager internally retries on
+                // retriable errors, so at this point we assume this is non-retriable, but
+                // proceed with the revocation anyway).
+                log.debug("Commit request before revocation failed with non-retriable error. Will" +
```

Comment:
Should we log this as an error?

Reply:
Yes, done. It is actually the level used for this in the legacy coordinator.